src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,236 @@
+import { HomeAssistant } from "../../../types.js";
+import { LovelaceCard, LovelaceConfig } from "../types.js";
+import {
+  clearItems,
+  fetchItems,
+  addItem,
+  completeItem,
+  saveEdit,
+  deleteItem,
+} from "../../../data/shopping_list";
+import { TemplateResult } from "lit-html";
+
+interface Config extends LovelaceConfig {
+  title?: string;
+}
+
+export interface ShoppingListItem {
+  id: number;
+  name: string;
+  complete: boolean;
+}
+
+class HuiShoppingListCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  protected _config?: Config;
+  private _items?: ShoppingListItem[];
+  private _unsubEvents?: () => Promise<void>;
+
+  static get properties() {
+    return {
+      _config: {},
+      _items: {},
+    };
+  }
+
+  public getCardSize(): number {
+    return (
+      (this._config ? (this._config.title ? 1 : 0) : 0) +
+      (this._items ? this._items.length : 3)
+    );
+  }
+
+  public setConfig(config: Config): void {
+    this._config = config;
+    this._items = [];
+    this._fetchData();
+  }
+
+  public connectedCallback() {
+    super.connectedCallback();
+
+    if (this.hass) {
+      this.hass.connection
+        .subscribeEvents(() => this._fetchData(), "shopping_list_updated")
+        .then(
+          (unsub: () => Promise<void>): void => {
+            this._unsubEvents = unsub;
+          }
+        );
+      this._fetchData();
+    }
+  }
+
+  public disconnectedCallback() {
+    super.disconnectedCallback();
+
+    if (this._unsubEvents) {
+      this._unsubEvents();
+    }
+  }
+
+  protected render(): TemplateResult {
+    if (!this._config || !this.hass) {
+      return html``;
+    }
+
+    return html`
+      ${this.renderStyle()}
+      <ha-card .header="${this._config.title}">
+        <div class="addRow">
+          <paper-input
+            id='addBox'
+            placeholder='${this.localize("ui.card.shopping-list.add_item")}'
+            @keydown=${this._addKeyPress}
+            no-label-float
+          ></paper-input>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
+            @click=${this._addItem}
+          ></paper-icon-button>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:notification-clear-all"
+            @click=${this._clearItems}
+          ></paper-icon-button>
+        </div>
+
+      ${repeat(
+        this._items!,
+        (item) => html`
+        <div class="editRow">
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
+            .itemId=${item.id}
+            @click=${this._completeItem}
+            tabindex='0'
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input
+              no-label-float
+              .value='${item.name}'
+              .itemId=${item.id}
+              @change=${this._saveEdit}

```

Comment:
This will fire A LOT of events. Every key press will fire a change event, that's then being sent to the server, changing an item generates an event, event is acted upon by this card by fetching items again. Then the user presses another key…

Reply:
It only fires when we lose focus